### PR TITLE
ClearlyDefined ScanResultStorage

### PR DESCRIPTION
**Pull request by oheger-bosch**
This PR provides an initial implementation of a (read-only) ScanResultsStorage based on ClearlyDefined (refer to issue #2252).

The coordinates of the package to be looked up in the storage are transformed to a form understandable by ClearlyDefined. This transformation is based on already existing functionality that was originally located in the _analyzer_ module, but has been moved to _model_, so that it can be referenced by both the _analyzer_ and the _scanner_ modules.

Interaction with ClearlyDefined is done via the pre-existing _ClearlyDefinedService_, which has been slightly extended to support a few more API calls. The storage implementation just queries whether harvested data for the packages in question is available and does not trigger a harvest operation itself.

The new _ClearlyDefinedStorage_ class is not yet integrated into the ORT configuration, so it cannot be used so far; this will be part of a follow-up pull request.

To test the storage implementation, tests have been written based on the [Wiremock](http://wiremock.org/) framework, to simulate the ClearlyDefined server. If desired, I can take a stab at an alternative testing framework.

### CODE REVIEW

File: clearly-defined/src/main/kotlin/ClearlyDefinedService.kt [hunk 1]
```diff
@@ -30,6 +30,7 @@ import java.io.File
 import java.net.URL
 
 import okhttp3.OkHttpClient
+import okhttp3.ResponseBody
```
Comment (sschuberth) [inline clearly-defined/src/main/kotlin/ClearlyDefinedService.kt:33]: Commit message nit: s/scancode/ScanCode/

File: scanner/src/main/kotlin/scanners/ScanCode.kt [hunk 2]
```diff
@@ -83,11 +72,12 @@ class ScanCode(
     name: String,
     config: ScannerConfiguration
 ) : LocalScanner(name, config) {
-    class Factory : AbstractScannerFactory<ScanCode>("ScanCode") {
+    class Factory : AbstractScannerFactory<ScanCode>(SCANNER_NAME) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCode.kt:75]: Commit message: Please stick to the imperative, "[...] has been added, too." ->"[...] is added, too."

File: scanner/src/main/kotlin/scanners/ScanCode.kt [hunk 3]
```diff
@@ -83,11 +72,12 @@ class ScanCode(
...
         override fun create(config: ScannerConfiguration) = ScanCode(scannerName, config)
     }
 
     companion object {
+        const val SCANNER_NAME = "ScanCode"
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCode.kt:80]: Nit: Blank line after this one to separate from `private` constants.

File: scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt [hunk 4]
```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (C) 2019 Bosch Software Innovations GmbH
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt:2]: This should be "Copyright (C) 2020 Bosch.IO GmbH", but before that please also add "Copyright (C) 2017-2019 HERE Europe B.V." as this is mostly moved code.

File: scanner/src/test/kotlin/ScanCodeTest.kt [hunk 5]
```diff
@@ -19,11 +19,14 @@
 
 package org.ossreviewtoolkit.scanner
 
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
```
Comment (sschuberth) [inline scanner/src/test/kotlin/ScanCodeTest.kt:23]: Nit: Blank line after this one.

File: analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt [hunk 6]
```diff
@@ -23,103 +23,22 @@ import java.io.IOException
...
-import org.ossreviewtoolkit.clearlydefined.Provider
-import org.ossreviewtoolkit.downloader.VcsHost
 import org.ossreviewtoolkit.model.Hash
 import org.ossreviewtoolkit.model.Identifier
 import org.ossreviewtoolkit.model.PackageCuration
 import org.ossreviewtoolkit.model.PackageCurationData
 import org.ossreviewtoolkit.model.RemoteArtifact
 import org.ossreviewtoolkit.model.VcsInfoCurationData
 import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.model.utils.toClearlyDefinedTypeAndProvider
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt:36]: Commit message nit: No trailing dot on the title.

File: model/build.gradle.kts [hunk 7]
```diff
@@ -34,6 +34,7 @@ plugins {
 dependencies {
     api(project(":spdx-utils"))
     api(project(":utils"))
+    api(project(":clearly-defined"))
```
Comment (sschuberth) [inline model/build.gradle.kts:37]: Please sort these alphabetically.

File: gradle.properties [hunk 8]
```diff
@@ -59,6 +59,7 @@ svnkitVersion = 1.10.1
 toml4jVersion = 0.7.2
 xalanVersion = 2.7.2
 xzVersion = 1.8
+wiremockVersion = 2.27.2
```
Comment (sschuberth) [inline gradle.properties:62]: Commit message nit: No trailing dot on the title.

File: scanner/build.gradle.kts [hunk 9]
```diff
@@ -33,10 +34,12 @@ dependencies {
 
     implementation(project(":downloader"))
     implementation(project(":utils"))
+    implementation(project(":clearly-defined"))
```
Comment (sschuberth) [inline scanner/build.gradle.kts:37]: Please sort entries alphabetically.

File: scanner/build.gradle.kts [hunk 10]
```diff
@@ -33,10 +34,12 @@ dependencies {
...
 
     implementation("com.fasterxml.jackson.module:jackson-module-kotlin:$jacksonVersion")
     implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:$kotlinxCoroutinesVersion")
     implementation("org.postgresql:postgresql:$postgresVersion")
 
+    testImplementation("com.github.tomakehurst:wiremock:$wiremockVersion")
```
Comment (sschuberth) [inline scanner/build.gradle.kts:43]: Please separate different Gradle configurations by a blank line.

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+ * A storage implementation that tries to download ScanCode results from ClearlyDefined.
+ *
+ * This storage implementation maps the requested package to coordinates used by ClearlyDefined. It then tries
+ * to find a harvested ScanCode result for these coordinates using the ClearlyDefined REST API.
+ */
+class ClearlyDefinedStorage(
+    /** The configuration for this storage implementation. */
+    val configuration: ClearlyDefinedStorageConfiguration
+) : ScanResultsStorage() {
+    companion object {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:60]: We should probably consider making all `companion object` members top-level members in the file instead to reduce nesting for better readability.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+import org.ossreviewtoolkit.model.ScannerDetails
+import org.ossreviewtoolkit.model.Success
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.model.config.ClearlyDefinedStorageConfiguration
+
+class ClearlyDefinedStorageTest : WordSpec({
+    lateinit var wiremock: WireMockServer
+
+    beforeSpec {
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:62]: Nice! I wasn't even aware of this syntax, and always did explicitly override the functions.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+            val storage = ClearlyDefinedStorage(ClearlyDefinedStorageConfiguration((serverUrl)))
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Failure ->
+                    result.error shouldContain "ConnectException"
+                else -> fail("Unexpected result: $result")
+            }
+        }
+    }
+}) {
+    companion object {
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:257]: We should probably consider making all `companion object` members top-level members in the file instead to reduce nesting for better readability.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/scanners/ScanCode.kt (hunk 2 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/ScanCode.kt:75]: All commit messages have been reworked.

File: scanner/src/main/kotlin/scanners/ScanCode.kt (hunk 3 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/ScanCode.kt:80]: Done

File: scanner/src/test/kotlin/ScanCodeTest.kt (hunk 5 above)
Comment (oheger-bosch) [reply scanner/src/test/kotlin/ScanCodeTest.kt:23]: Done

File: model/build.gradle.kts (hunk 7 above)
Comment (oheger-bosch) [reply model/build.gradle.kts]: Done

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt (hunk 11 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt]: Removed companion objects in ClearlyDefinedStorage and test class and moved declarations to top-level.

File: scanner/src/main/kotlin/scanners/ScanCode.kt [hunk 14]
```diff
@@ -83,11 +72,13 @@ class ScanCode(
     name: String,
     config: ScannerConfiguration
 ) : LocalScanner(name, config) {
```
Comment (fviernau) [inline scanner/src/main/kotlin/scanners/ScanCode.kt:74]: The tests could also be moved to their own file

File: analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt [hunk 15]
```diff
@@ -23,103 +23,22 @@ import java.io.IOException
 
 import org.ossreviewtoolkit.analyzer.PackageCurationProvider
 import org.ossreviewtoolkit.clearlydefined.ClearlyDefinedService
```
Comment (fviernau) [inline analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt:25]: commit message: Please use present tense: "were moved" -> "are moved"

File: gradle.properties (hunk 8 above)
Comment (fviernau) [inline gradle.properties:62]: Please adhere to ordering of the version variables.

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+     * well, the marshalled body of the request is returned; otherwise, the function throws an exception.
+     */
+    private fun <T> execute(call: Call<T>): T {
+        val req = "${call.request().method} ${call.request().url}"
+        log.debug("Executing $req")
+        val response = call.execute()
+        log.debug("${response.code()} ${response.message()}")
+        if (response.isSuccessful) {
+            return response.body()!!
+        } else {
```
Comment (fviernau) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:202]: The ` else` could be removed. If you should do that an empty line above log.error would be nice.

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    /**
+     * Execute an HTTP request specified by the given [call]. The response status is checked. If everything went
+     * well, the marshalled body of the request is returned; otherwise, the function throws an exception.
+     */
+    private fun <T> execute(call: Call<T>): T {
+        val req = "${call.request().method} ${call.request().url}"
+        log.debug("Executing $req")
+        val response = call.execute()
+        log.debug("${response.code()} ${response.message()}")
+        if (response.isSuccessful) {
```
Comment (fviernau) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:200]: Please add an empty line above.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+
+    afterSpec {
+        wiremock.stop()
+    }
+
+    beforeTest {
+        wiremock.resetAll()
+    }
+
+    "ClearlyDefinedStorage" should {
```
Comment (fviernau) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:214]: Please add an empty line before the first assertion of each test.
...to get a clearer view on the 'arrange', 'act' 'assert' blocks.

File: scanner/src/main/kotlin/scanners/ScanCode.kt (hunk 14 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/ScanCode.kt:74]: Moved the tests affected to a new test class ScanCodeResultParserTest.

File: scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+/**
+ * Generate a ScannerDetails object from the given [result] node, which structure depends on the current ScanCode
+ * version. The node names to check are specified via [optionsNode], and [versionNode].
+ */
+private fun generateScannerDetailsFromNode(result: JsonNode, optionsNode: String, versionNode: String):
+        ScannerDetails {
+    val config = generateScannerOptions(result[optionsNode])
+    val version = result[versionNode]?.asText() ?: ""
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt:208]: Prefer: `result[versionNode].textValueOrEmpty()`

File: scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+
+    return options?.let {
+        val optionList = it.fieldNames().asSequence().fold(mutableListOf<String>()) { list, opt ->
+            addValues(list, it[opt], opt)
+            list
+        }
+        optionList.joinToString(separator = " ")
+    } ?: ""
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt:235]: Prefer: `.orEmpty()`

File: scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+    return ScannerDetails(ScanCode.SCANNER_NAME, version, config)
+}
+
+/**
+ * Convert the JSON node with ScanCode [options] to a string that corresponds to the options as they have been
+ * passed on the command line.
+ */
+private fun generateScannerOptions(options: JsonNode?): String {
+    fun addValues(list: MutableList<String>, node: JsonNode, key: String) {
+        if (node.size() == 0) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt:218]: Prefer: `.isEmpty`

File: scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+            list.add(key)
+            list.add(node.asText())
+        } else {
+            (0 until node.size()).forEach { idx ->
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/ScanCodeResultParser.kt:222]: Prefer: `node.forEachIndexed { index, _ -> ... }`

File: scanner/src/test/kotlin/ScanCodeTest.kt [hunk 23]
```diff
@@ -832,8 +167,8 @@ class ScanCodeTest : WordSpec({
         "contain the default values if the scanner configuration is empty" {
             scanner.commandLineOptions.joinToString(" ") shouldMatch
                     "--copyright --license --ignore \\*$ORT_REPO_CONFIG_FILENAME --info --strip-root --timeout 300 " +
-                            "--ignore HERE_NOTICE --ignore META-INF/DEPENDENCIES --processes \\d+ --license-diag " +
-                            "--verbose"
+                    "--ignore HERE_NOTICE --ignore META-INF/DEPENDENCIES --processes \\d+ --license-diag " +
+                    "--verbose"
```
Comment (sschuberth) [inline scanner/src/test/kotlin/ScanCodeTest.kt:171]: Nit: This indentation change should be reverted.

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+import org.ossreviewtoolkit.model.utils.toClearlyDefinedCoordinates
+import org.ossreviewtoolkit.model.utils.toClearlyDefinedSourceLocation
+import org.ossreviewtoolkit.scanner.ScanResultsStorage
+import org.ossreviewtoolkit.scanner.scanners.generateScannerDetails
+import org.ossreviewtoolkit.scanner.scanners.generateSummary
+import org.ossreviewtoolkit.utils.log
+
+import retrofit2.Call
+
+/** The name used by ClearlyDefined for the ScanCode tool. */
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:50]: So far we haven't used KDoc one-liner comments like these; I wasn't even aware they're supported. I somewhat like them for their compactness, but I'm wondering what's your rule to use one-liner comments, as e.g. in line 54 you didn't. Could you please elaborate?

Comment (sschuberth) [reply scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:50]: Do you maybe limit the use to properties?

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+     */
+    private fun readPackageFromClearlyDefined(
+        id: Identifier,
+        vcs: VcsInfo?,
+        sourceArtifact: RemoteArtifact?
+    ): Result<ScanResultContainer> =
+        try {
+            readFromClearlyDefined(id, packageCoordinates(id, vcs, sourceArtifact))
+        } catch (e: IllegalArgumentException) {
+            log.warn("Could not obtain ClearlyDefined coordinates for package $id", e)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:129]: Please use `'${id.toCoordinates()}'` when printing `Identifier`s. Also: Missing dot at end of sentence 😀

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+     * Try to obtain a [ScanResult] produced by ScanCode from ClearlyDefined for the package with the given [id],
+     * using the given [coordinates] for looking up the data. These may not necessarily be equivalent to the
+     * identifier, as we try to lookup source code results if a GitHub repository is known.
+     */
+    private fun readFromClearlyDefined(
+        id: Identifier,
+        coordinates: ClearlyDefinedService.Coordinates
+    ): Result<ScanResultContainer> {
+        val startTime = Instant.now()
+        log.info("Looking up results for ${id.toCoordinates()}.")
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:143]: Missing single quotes around `${id.toCoordinates()}`.

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+
+        return try {
+            val packageUrl = coordinates.toString()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:146]: Maybe inline this variable as it's only used once?

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+                    coordinates.name,
+                    coordinates.revision.orEmpty()
+                )
+            )
+            val scanCodeVersion = findScanCodeVersion(tools, packageUrl)
+            scanCodeVersion?.let { version ->
+                loadScanCodeResults(id, coordinates, version, startTime)
+            } ?: emptyResult(id)
+        } catch (e: IOException) {
+            log.error("Error when reading results for package $id from ClearlyDefined", e)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:161]: Please use `'${id.toCoordinates()}'` when printing `Identifier`s. Also: Missing dot at end of sentence 😀

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+        version: String,
+        startTime: Instant
+    ): Result<ScanResultContainer> {
+        val toolResponse = execute(
+            clearlyDefinedService.harvestToolData(
+                coordinates.type, coordinates.provider, coordinates.namespace.orEmpty(), coordinates.name,
+                coordinates.revision.orEmpty(), TOOL_SCAN_CODE, version
+            )
+        )
+        return toolResponse.use {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:182]: Nit: Blank line before this one.

File: scanner/src/test/assets/__files/clearlydefined_scancode.json [hunk 30]
```diff
@@ -0,0 +1,853 @@
+{
```
Comment (sschuberth) [inline scanner/src/test/assets/__files/clearlydefined_scancode.json:1]: Where does the `__files` convention in the path come from? Should we maybe be more specific and replace it with "wiremock(s)"? Or slightly less specific and call it "response(s)"?

File: scanner/src/test/assets/__files/clearlydefined_scancode.json [hunk 31]
```diff
@@ -0,0 +1,853 @@
...
+        "is_generated": false,
+        "is_license_text": false,
+        "files_count": 0,
+        "dirs_count": 0,
+        "size_count": 0,
+        "scan_errors": []
+      }
+    ]
+  }
+}
```
Comment (sschuberth) [inline scanner/src/test/assets/__files/clearlydefined_scancode.json:853]: Nit: Missing newline at end of file.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 32]
```diff
@@ -0,0 +1,405 @@
...
+private val MAX_TIME_DELTA = Duration.ofSeconds(30)
+
+/** The ClearlyDefined URL referencing the test package. */
+private const val PACKAGE_URL = "maven/mavencentral/$NAMESPACE/$NAME/$VERSION"
+
+/** Path to a file contained in the test ClearlyDefined result. */
+private const val TEST_PATH =
+    "src/main/java/org/apache/commons/configuration2/tree/DefaultExpressionEngine.java"
+
+/** A test identifier. */
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:79]: Nit: I think it's fine to not overdo it and omit comments that basically just repeat the variable name.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 33]
```diff
@@ -0,0 +1,405 @@
...
+private val TEST_IDENTIFIER =
+    Identifier(
+        type = PACKAGE_TYPE,
+        namespace = NAMESPACE,
+        name = NAME,
+        version = VERSION
+    )
+
+/** A test package. */
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:88]: See above.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 34]
```diff
@@ -0,0 +1,405 @@
...
+    }
+
+/**
+ * Check that the given [result] for package [expId] does not contain any data.
+ */
+private fun assertEmptyResult(result: Result<ScanResultContainer>, expId: Identifier = TEST_IDENTIFIER) {
+    when (result) {
+        is Success -> {
+            result.result.id shouldBe expId
+            result.result.results shouldHaveSize 0
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:176]: Prefer: `should beEmpty()`

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 35]
```diff
@@ -0,0 +1,405 @@
...
+                get(anyUrl())
+                    .willReturn(aResponse().withStatus(500))
+            )
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Success -> fail("Expected failure, but got $result")
+                is Failure ->
+                    result.error shouldContain "IOException"
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:280]: Nit: This looks like it should still fit into the previous line.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 36]
```diff
@@ -0,0 +1,405 @@
...
+                        aResponse().withStatus(200)
+                            .withBody("This is not a JSON response")
+                    )
+            )
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Failure ->
+                    result.error shouldContain "JsonParseException"
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:370]: Nit: This looks like it should still fit into the previous line.

File: scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt [hunk 37]
```diff
@@ -0,0 +1,405 @@
...
+        "handle a failure to connect to the server" {
+            // find a port on which no service is running
+            val port = ServerSocket(0).use { it.localPort }
+            val serverUrl = "http://localhost:$port"
+
+            val storage = ClearlyDefinedStorage(ClearlyDefinedStorageConfiguration((serverUrl)))
+
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Failure ->
+                    result.error shouldContain "ConnectException"
```
Comment (sschuberth) [inline scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt:400]: Nit: This looks like it should still fit into the previous line.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt (hunk 24 above)
Comment (oheger-bosch) [reply scanner/src/main/kotlin/storages/ClearlyDefinedStorage.kt:50]: And I was not aware of a potential difference to Javadoc ;-)

I use this style for properties if the description is short enough.  For functions I am used to multi-line comments from my Java background. Here comments for methods normally span multiple lines as they contain tags for parameters and return values.

File: scanner/src/test/assets/__files/clearlydefined_scancode.json (hunk 30 above)
Comment (oheger-bosch) [reply scanner/src/test/assets/__files/clearlydefined_scancode.json:1]: This is unfortunately a convention of the WireMock library. You can configure the root path of your test files, but it insists on appending a "__files" subfolder. I have found no way to change this.

Comment (sschuberth) [reply scanner/src/test/assets/__files/clearlydefined_scancode.json:1]: Good to know. How about add a very brief `README.md` then to the `__files` directory, explaining exactly this WireMock convention? So people just browsing the files do not wonder like I did.

File: scanner/src/test/assets/__files/README.md [hunk 38]
```diff
@@ -0,0 +1,5 @@
+This directory contains test files for tests using the [WireMock](http://wiremock.org/) framework.
+
+It is a convention of the test framework that the files to be served by the mock server must be located in a
+folder named `__files`.
+
```
Comment (sschuberth) [inline scanner/src/test/assets/__files/README.md:5]: There's a single trailing space in this line which should be removed 😁

**Review by sschuberth — APPROVED** [review]

**Pull request merged**